system/lib/wasmfs/syscalls.cpp

```diff
@@ -23,6 +25,51 @@ extern "C" {
 
 using namespace wasmfs;
 
+// Copy the file specified by the pathname into JS.
+// Return a pointer to the JS buffer in HEAPU8.
+// The buffer will also contain the file length.
+// Caller must free the returned pointer.
+void* emscripten_wasmfs_read_file(char* path) {
+  struct stat file;
+  int err = 0;
+  err = stat(path, &file);
+  if (err < 0) {
+    emscripten_console_error("Fatal error in FS.readFile");
+    abort();
+  }
+
+  // file.st_size could exceed uint32_t in wasm64.
+  if (file.st_size > std::numeric_limits<uint32_t>::max()) {
+    emscripten_console_error(
+      "Fatal error in FS.readFile: file size has exceeded uint32_t");
+    abort();
+  }
+
+  // Result will return the a pointer to a buffer with the file length in the
+  // first 4 bytes. The remaining bytes will contain the buffer contents. This
+  // allows the caller to use HEAPU8.subarray(buf + 4, buf + 4 + length).
+  uint32_t size = file.st_size;
```

Comment:
I think this will silently truncate if `st_size` is larger than 32 bits? We should check and abort if it is too big. Or, we could use a 64-bit size here, because the size of files is in fact 64-bit?

Reply:
Let's use an `off_t` instead, that makes more sense.